app/features/worksheet_generator/tests/test_tools.py

```diff
@@ -0,0 +1,280 @@
+import os
+from dotenv import load_dotenv, find_dotenv
+import logging
+import pytest
+from app.features.worksheet_generator.tools import RAGpipeline, TextOrPDFLoader, UploadPDFLoader, read_text_file
+from app.features.worksheet_generator.tools import RAGRunnable
+from app.features.worksheet_generator.tools import BytesFilePDFLoader
+from unittest.mock import MagicMock, patch
+from io import BytesIO
+from fastapi import UploadFile
+# from PyPDF2 import PdfReader
+
+load_dotenv(find_dotenv())
+
+# default logging setting:
+logging.basicConfig(level=logging.INFO)
+
+# Fixture Section
+
+    # Create a mock file, called sample.txt.
+@pytest.fixture
+def sample_text_file(tmpdir):
+    file_path = tmpdir.join("sample.txt")
+    file_path.write("This is a test file.")
+    return file_path
+
+@pytest.fixture
+    # Creating a mock UploadFile object
+def mock_upload_file():
+
+    upload_file = MagicMock(spec=UploadFile)
+    upload_file.filename = "sample.pdf"
+    
+    pdf_content = b"%PDF-1.4\n\n1 0 obj\n<< /Type /Catalog /Pages 3 0 R >>\nendobj\n\n2 0 obj\n<< /Type /Page /Parent 3 0 R /Contents 5 0 R >>\nendobj\n\n3 0 obj\n<< /Type /Pages /Kids [2 0 R] /Count 1 >>\nendobj\n\n4 0 obj\n<< /Title (Sample Document) /Creator (Me) /Producer (My PDF Library 1.0) >>\nendobj\n\n5 0 obj\n<< /Length 20 >>\nstream\n0.2 0.2 0.2 rg\nBT\n/F1 12 Tf\n72 712 Td\n(Hello World) Tj\nET\nendstream\nendobj\n\n6 0 obj\n<< /Type /Font /Subtype /Type1 /Name /F1 /BaseFont /Times-Roman >>\nendobj\n\nxref\n0 7\n0000000000 65535 f \n0000000009 00000 n \n0000000054 00000 n \n0000000102 00000 n \n0000000174 00000 n \n0000000275 00000 n \n0000000354 00000 n \ntrailer\n<< /Size 7 /Root 1 0 R >>\nstartxref\n460\n%%EOF\n"
```

Comment:
Including PDF content directly in the code makes it unnecessarily cluttered and is not considered a best practice.